juju/juju
# Model Cache Mark and Sweep

## Description of change

This patch adds mark/sweep logic to the cache resident manager.
- Each resident is provided a removal message upon instantiation.
- `Mark` is invoked when the mode cache worker's multi-watcher throws an error and restarts.
- After the first deltas are processed following such a restart, cache `Sweep` is invoked.
- Any residents still stale send their removal messages to the cache controller's change channel causing them to be evicted from the cache.

## QA steps

Unit tests verify the new functionality.

QA for regression is the same as for https://github.com/juju/juju/pull/10062.

## Documentation changes

None.

## Bug reference

N/A
